### PR TITLE
Virtualized unreadmarker

### PR DESCRIPTION
**Pull request by Trolldemorted**
Basic unread markers for the virtualized collection! Unread state is currently only changed by composing a message and receiving a sync message. Will refine this in a few days, but that involves reacting to scrolling and should go into a separate PR.

Note: This PR includes a database migration, and 2 minor unrelated bugfixes.

cc @golf1052

### CODE REVIEW

**Comment by Trolldemorted** [comment]
rebased to the current master

File: Signal-Windows/Controls/Conversation.xaml [hunk 1]
```diff
@@ -57,7 +36,16 @@
...
         </Border>
-        <ListBox Grid.Row="1" Name="ConversationItemsControl" VirtualizingStackPanel.VirtualizationMode="Recycling" Background="White" ScrollViewer.VerticalScrollBarVisibility="Visible" Padding="0 0 15 0" ItemContainerStyleSelector="{StaticResource MessageDataStyleSelector}" />
+        <ListView Grid.Row="1" Name="ConversationItemsControl" VirtualizingStackPanel.VirtualizationMode="Recycling" Background="White" ScrollViewer.VerticalScrollBarVisibility="Visible" Padding="0 0 15 0" ItemTemplateSelector="{StaticResource MessageDataTemplateSelector}" SelectionMode="None">
+            <ListView.ItemContainerStyle>
+                <Style TargetType="ListViewItem">
+                    <Setter Property="HorizontalContentAlignment" Value="Stretch" />
+                </Style>
+            </ListView.ItemContainerStyle>
+            <ListView.Style>
+                <Style TargetType="ListView" />
```
Comment (golf1052) [inline Signal-Windows/Controls/Conversation.xaml]: Not sure what this is for.

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs [hunk 2]
```diff
@@ -1,142 +1,230 @@
...
+    }
+
+    public class SignalUnreadMarker
+    {
+        public string Text = "";
+    }
+
+    public class VirtualizedCollection : IList, INotifyCollectionChanged
+    {
+        private readonly static int PAGE_SIZE = 50;
```
Comment (golf1052) [inline Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: const

File: Signal-Windows/Controls/Conversation.xaml.cs [hunk 3]
```diff
@@ -183,21 +180,27 @@ public void UpdateMessageBox(SignalMessage updatedMessage)
...
                 }
             }
         }
 
         public void Append(SignalMessageContainer sm, bool forceScroll)
         {
-            Collection.Add(sm);
-            if (forceScroll || true) //TODO
+            var sourcePanel = (ItemsStackPanel)ConversationItemsControl.ItemsPanelRoot;
+            bool bottom = sourcePanel.LastVisibleIndex == Collection.Count - 2; /* -2 because we already incremented Count */
```
Comment (golf1052) [inline Signal-Windows/Controls/Conversation.xaml.cs]: Not sure what the comment means, where did we increment Count?

File: Signal-Windows/Controls/UnreadMarker.xaml.cs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (golf1052) [inline Signal-Windows/Controls/UnreadMarker.xaml.cs]: Unneeded

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs [hunk 5]
```diff
@@ -1,142 +1,230 @@
-﻿using Signal_Windows.Models;
```
Comment (golf1052) [inline Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: Looks like the line endings changed on the file, makes it hard to view the diff :|

**Comment by golf1052** [comment]
Pretty much looks good to me. Just a couple of questions.

File: Signal-Windows/Controls/Conversation.xaml (hunk 1 above)
Comment (Trolldemorted) [reply Signal-Windows/Controls/Conversation.xaml]: you are right, this does nothing and can be removed

File: Signal-Windows/Controls/Conversation.xaml.cs (hunk 3 above)
Comment (Trolldemorted) [reply Signal-Windows/Controls/Conversation.xaml.cs]: When we arrive here, the message has already been added to the database, so the collection's count has already been increased [here](https://github.com/signal-csharp/Signal-Windows/blob/virtualized_unreadmarker/Signal-Windows/ViewModels/MainPageViewModel.cs#L369), [here ](https://github.com/signal-csharp/Signal-Windows/blob/virtualized_unreadmarker/Signal-Windows/ViewModels/MainPageViewModel.cs#L425) or [here](https://github.com/signal-csharp/Signal-Windows/blob/virtualized_unreadmarker/Signal-Windows/ViewModels/MainPageViewModel.cs#L290).

We are at the bottom if the previous message's index is equal to the LastVisibleIndex.

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs (hunk 5 above)
Comment (Trolldemorted) [reply Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: file says it is now `ASCII text, with CRLF line terminators`

Maybe d2086564522793742710f5406d7b7ea15ab8adc9 committed it with LF feeds into master?

File: Signal-Windows/Controls/VirtualizedMessagesCollection.cs (hunk 2 above)
Comment (Trolldemorted) [reply Signal-Windows/Controls/VirtualizedMessagesCollection.cs]: 👍

**Pull request merged**